File: framework/java/test/04/Leet.java
import soda.unittest.*;

import java.util.*;
import java.util.function.Supplier;
import java.util.stream.*;

import soda.leetcode.*;
import soda.unittest.work.TestWork;

import static soda.unittest.LoggerHelper.logger;

class TopVotedCandidate {

    int N = 0;
    int[] times;
    int[] winner;

    public TopVotedCandidate(int[] persons, int[] times) {
        N = persons.length;
        this.times = times;
        winner = new int[N];

        int[] counter = new int[N+1];
        int win = 0;
        for (int i = 0; i < N; ++i) {
            if (++counter[persons[i]] >= counter[win]) {
                win = persons[i];
            }
            winner[i] = win;
        }
    }
    
    public int q(int t) {
        if (t >= times[N-1]) {
            return winner[N-1];
        }
        int low = 0, high = N-1;
        while (low < high) {
            int mid = (low + high) / 2;
            if (t <= times[mid]) {
                high = mid;
            } else {
                low = mid + 1;
            }
        }
        return t == times[low] ? winner[low] : winner[low-1];
    }
}

class Params {
    public int[] persons;
    public int[] times;
    public int[] qs;
}

class Solution {
    public List<Integer> doTest(String[] commands, Params params) {
        List<Integer> res = new ArrayList<>();
        TopVotedCandidate tvc = null;
        for (int i = 0; i < commands.length; ++i) {
            if (commands[i].equals("TopVotedCandidate")) {
                tvc = new TopVotedCandidate(params.persons, params.times);
                res.add(null);
            } else {
                res.add(tvc.q(params.qs[i]));
            }
        }
        return res;
    }
}

public class Leet implements Supplier<TestWork> {

    @Override
    public TestWork get() {
        var work = new TestWork(new Solution(), "doTest");
        // work.setValidator((e, r) -> {...});
        work.setCompareSerial(true);
        work.setArgumentParser(1, (List<Object> v) -> {
            Params params = new Params();
            var p0 = (List<List<Integer>>) v.get(0);
            params.persons = DataUtils.toIntArray(p0.get(0));
            params.times = DataUtils.toIntArray(p0.get(1));
            params.qs = new int[params.persons.length];
            for (int i = 1; i < v.size(); ++i) {
                List<Integer> p = (List<Integer>) v.get(i);
                params.qs[i] = p.get(0);
            }
            return params;
        });
        // work.setResultParser(r -> { ... });
        // work.setResultSerializer(r -> {...});
        return work;
    }

    public static void main(String[] args) throws Exception {
        new Leet().get().run();
    }

}

File: framework/test/04/leet.cpp
//#include <bits/stdc++.h>
#include <vector>
#include <optional>
#include <iostream>
#include <string>

// #include "soda/leetcode/leetcode.h"
#include "soda/unittest/work.h"

using namespace std;
using namespace soda::leetcode;
using namespace soda::unittest;

// step [1]: implement class Solution
// class Solution {};
class TopVotedCandidate {
    int N;
    vector<int> times;
    vector<int> winner;
public:
    TopVotedCandidate(vector<int>& persons, vector<int>& times):
        N(persons.size()), times(times), winner(persons.size())
    {
        vector<int> counter(N+1, 0);
        int win = 0;
        for (int i = 0; i < N; ++i) {
            if (++counter[persons[i]] >= counter[win]) {
                win = persons[i];
            }
            winner[i] = win;
        }
    }
    
    int q(int t) {
        if (t >= times.back()) {
            return winner[N-1];
        }
        int low = 0, high = N-1;
        while (low < high) {
            int mid = (low + high) / 2;
            if (t <= times[mid]) {
                high = mid;
            } else {
                low = mid + 1;
            }
        }
        return t == times[low] ? winner[low] : winner[low-1];
    }
};

namespace {
    const auto __ = []() {
        ios_base::sync_with_stdio(false);
        cin.tie(nullptr);
        return 0;
    }();
}

struct Params {
    vector<int> persons;
    vector<int> times;
    vector<int> qs;
};

class Solution {
public:
    vector<optional<int>> doTest(vector<string>& commands, Params& params) {
        vector<optional<int>> res;
        TopVotedCandidate* tvc = nullptr;
        for (int i = 0; i < commands.size(); ++i) {
            if (commands[i] == "TopVotedCandidate") {
                tvc = new TopVotedCandidate(params.persons, params.times);
                res.emplace_back();
            } else {
                res.emplace_back(tvc->q(params.qs[i]));
            }
        }
        delete tvc;
        return res;
    }
};

USE_CUSTOM_SERIAL(Params)

class ParamsParser : public TypedDataParser<Params> {
public:
    Params parse(JsonPointer jp) override {
        Params params;
        auto p0 = jp[0].get<vector<vector<int>>>();
        params.persons = p0[0];
        params.times = p0[1];
        params.qs.push_back(-1);
        for (int i = 1; i < jp.size(); ++i) {
            auto v = jp[i].get<vector<int>>();
            params.qs.push_back(v[0]);
        }
        return params;
    }
};

// [optional] use custom type with parser and serializer,
// and if you use this macro, may be you should define your own validator
// USE_CUSTOM_SERIALIZER(type)

// [optional] instantiate json access type. This will increase compile time
// #include "soda/unittest/json_access.h"
// SODA_JSON_ACCESS_TYPE(type)

int main()
{
    // [1] create by class member function
    Solution su;
    auto work = WorkFactory::create(su, &Solution::doTest);
    //
    // [2] or, create by ordinary function
    // auto work = WorkFactory::create(function);

    // work->setValidator(validate);
    work->setCompareSerial(true);
    work->setArgParser<1>(new ParamsParser);
    // work->setArgParser<0,from_type>(parse_func);
    // work->setResultParser<from_type>(parse_func);
    // work->setResultSerializer(serial_func);
    work->run();
    delete work;
    return 0;
}
